Strictly suppress `file_matcher` in attribute completions

Darshan808 opened the pull request:
### Fixes #15031 

### Description
The current logic only suppresses matchers if they are in the suppress set and there are matching results. As a result, `file_matcher` can still display file matches when attribute completion fails. This PR updates the logic to always suppress `file_matcher` during attribute completion, regardless of whether any attributes were successfully matched.

Darshan808 committed:
respect-file-matcher-suppress-in-attr-completions

Darshan808 committed:
simplify code

Darshan808 committed:
Merge branch 'main' into suppress-file-completion-enh

krassowski reviewed (changes requested):
This would break completion on `ls` automagic:

| Before | This PR as of now |
|--|--|
| <img width="1091" height="110" alt="Image" src="https://github.com/user-attachments/assets/5ebb35b6-39c9-4530-98f8-3d9265f4785b" /> | <img width="1091" height="201" alt="Image" src="https://github.com/user-attachments/assets/cfa3df5b-7745-4eb6-b540-dcfbe748203a" /> |

Can you add a test for the `ls` case?

Darshan808 committed:
fix CONTRIBUTING.md
COPYING.rst
IPython
LICENSE
MANIFEST.in
README.rst
SECURITY.md
__pycache__
_build_meta.py
codecov.yml
docs
examples
ipython.egg-info
long_description.rst
pyproject.toml
scripts
setup.cfg
setup.py
setupbase.py
tests
tools case

Darshan808 commented:
Oops, I didn't realize backticks inside double quotes get executed. My git commit -m "fix `ls .` case" ended up expanding the command.
Anyways, cases like `ls file.<tab>` still don't complete or suggest files with the same name but different extensions, since they're treated as attribute completion. This doesn't seem trivial to fix. Should we address it in this PR, or would it be better to open a separate issue?

krassowski commented:
> Anyways, cases like ls file.<tab> still don't complete or suggest files with the same name but different extensions, since they're treated as attribute completion. This doesn't seem trivial to fix. Should we address it in this PR, or would it be better to open a separate issue?

We should not introduce regressions in this PR. Let's take a step back and think about it differently.

I think there is only a limited set of commands which can be auto-evaluated without `!` which are stored in `get_ipython().alias_manager.aliases`. It feels that we should determine whether we are completing in such CLI alias, line  magics (`%`), or bang expressions (`!`). This would be similar to `_determine_completion_context` but I would probably add a separate method for simplicity. The logic for detecting if we are completing in cli-alias should be along the lines of
> does the line start with a string enumerated in `aliases` followed by a whitespace character and there is no Python variable which overshadows it (as in there is no `ls = 1` or similar defined)

Darshan808 committed:
check-for-cli-context

Darshan808 committed:
add-test-for-cli-context

Darshan808 committed:
debug-in-ci

Darshan808 committed:
use-custom-alias

Darshan808 committed:
debug

Darshan808 committed:
use test_alias as test is already on user_ns

krassowski commented on IPython/core/completer.py:
```diff
@@ -3651,10 +3695,25 @@ def _complete(self, *, cursor_line, cursor_pos, line_buffer=None, text=None,
+                            suppression_recommended
+                            and (suppression_config is not False)
+                        )
+                    ) and has_any_completions(result)
```
Looking at how this requires hard-coding matcher-specific behaviour (or breaking backward compatibility) I now think it would be better to go with your initial idea of short-circuiting in `file_matcher` itself.

Darshan808 replied:
That makes sense!

Darshan808 committed:
short-circuit-on-file-matcher-to-prevent-completions

Darshan808 committed:
handle-cli-aliases

krassowski committed:
Add more tests

krassowski reviewed (approved):
Thank you @Darshan808!
